id/id.go

```diff
@@ -0,0 +1,20 @@
+// Package id contains facilities for generating and validating random IDs
+// which are meant to be used as uploaded filenames.
+package id
+
+import (
+	"crypto/rand"
+	"math"
+	"math/big"
+)
+
+// Generate a random ID like "1njfizqgeukrq", "3l44ze7pf47fd", "35dgd4n5ryup", etc.
+func Generate() (id string, err error) {
+	n, err := rand.Int(rand.Reader, max)
+	if err != nil {
+		return "", err
+	}
+	return n.Text(36), nil
+}
+
+var max = new(big.Int).SetUint64(math.MaxUint64)
```

Comment:
It is indeed a pattern that happens in many Go projects, and it's counter to some other languages such as C/C++ where dependencies tend to be defined/declared first before things that use them.

In Go, package level declaration order largely doesn't matter, the compiler figures out the order based on which identifiers use/mention others (see https://golang.org/ref/spec#Package_initialization), so you can arrange them in whatever way is most readable.

It turns out that the best readability of code can be achieved, more often than not, by placing the highest level operations first, and the lower level helpers second. The reason it works well is because humans tend to read from top to bottom, and reading the higher level functions gives you _context_ for the lower level helpers. Once you see where the helpers are used, you can then understand why they're needed and what they do, when you see their implementation details later.

I was indeed making this declaration order intentionally and thoughtfully; it wasn't random.